compiler/src/iree/compiler/Codegen/Common/GPU/GPUReuseSharedMemoryAllocs.cpp

```diff
@@ -0,0 +1,227 @@
+      }
+      Operation *userAncestor = parentRegion.findAncestorOpInRegion(*user);
+      if (!begin && !end) {
+        begin = end = userAncestor;
+        continue;
+      }
+      if (dominanceInfo.dominates(end, userAncestor)) {
+        end = userAncestor;
+      }
+      if (dominanceInfo.dominates(userAncestor, begin)) {
+        begin = userAncestor;
+      }
+    }
+  }
+  return std::make_pair(begin, end);
+}
+
+/// Compute the liveness range within the single block of the `rootOp` for each
+/// shared memory allocation. For this to succeed the `rootOp` must have the
+/// following conditions:
+///
+/// 1. `rootOp` must have only a single region with a single block.
+/// 2. `rootOp` must not have control flow.
+/// 3. `rootOp` must have at least 2 shared memory allocations.
+static LogicalResult
+populateLivenessRanges(Operation *rootOp,
+                       DenseMap<Operation *, LivenessRange> &livenessMap,
+                       DominanceInfo &dominanceInfo) {
+  if (rootOp->getNumRegions() != 1) {
+    return failure();
+  }
+  if (!rootOp->getRegion(0).hasOneBlock()) {
+    return failure();
+  }
+  Block &rootBlock = rootOp->getRegion(0).getBlocks().front();
+  if (llvm::any_of(rootBlock.getOperations(), [](Operation &op) {
+        return isa_and_present<cf::ControlFlowDialect>(op.getDialect());
+      })) {
+    return failure();
+  }
+
+  // Collect all shared memory allocations.
+  SmallVector<memref::AllocOp> sharedMemAllocs;
+  for (auto alloc : rootBlock.getOps<memref::AllocOp>()) {
+    if (hasSharedMemoryAddressSpace(alloc.getType())) {
+      sharedMemAllocs.push_back(alloc);
+    }
+  }
+  if (sharedMemAllocs.size() < 2) {
+    return failure();
+  }
+
+  for (auto alloc : sharedMemAllocs) {
+    LivenessRange livenessRange =
+        getLivenessRange(alloc, dominanceInfo, rootOp->getRegion(0));
+    livenessMap.insert(std::make_pair(alloc, livenessRange));
+  }
+  return success();
+}
+
+/// Helper to determine if 2 liveness ranges overlap. The ranges overlap if
+/// the `begin` of one of the ranges falls on or between the `begin` and `end`
+/// of the other range.
+static bool livenessRangesOverlap(LivenessRange range, LivenessRange other,
+                                  DominanceInfo &dominanceInfo,
+                                  Block *entryBlock) {
+  return (dominanceInfo.dominates(range.first, other.first) &&
+          dominanceInfo.dominates(other.first, range.second)) ||
+         (dominanceInfo.dominates(other.first, range.first) &&
+          dominanceInfo.dominates(range.first, other.second));
+}
+
+/// Returns the sets of allocations that have overlapping liveness ranges.
+/// Each set will not have any overlapping liveness ranges with any other set.
+static SmallVector<AliasGroup>
+computeAliasGroups(DenseMap<Operation *, LivenessRange> livenessMap,
+                   DominanceInfo &dominanceInfo, Block *entryBlock) {
+  SmallVector<AliasGroup> aliasGroups;
+  for (auto val : livenessMap) {
```

Comment:
nit: no auto

Reply:
The type name is quite long since it is a DenseMap pair, and the types of the `first` and `second` element of the pair are shown right below. Is it okay to keep this one auto?

Reply:
Maybe use structured bindings instead `auto [key, val] : livenessMaps`, except with better names